noir-projects/noir-contracts/contracts/nft_contract/src/main.nr

```diff
@@ -0,0 +1,191 @@
+        assert(!admin.is_zero(), "invalid admin");
+        storage.admin.write(admin);
+        storage.minters.at(admin).write(true);
+        storage.name.initialize(FieldCompressedString::from_string(name));
+        storage.symbol.initialize(FieldCompressedString::from_string(symbol));
+    }
+
+    #[aztec(public)]
+    fn set_admin(new_admin: AztecAddress) {
+        assert(storage.admin.read().eq(context.msg_sender()), "caller is not admin");
+        storage.admin.write(new_admin);
+    }
+
+    #[aztec(public)]
+    fn set_minter(minter: AztecAddress, approve: bool) {
+        assert(storage.admin.read().eq(context.msg_sender()), "caller is not admin");
+        storage.minters.at(minter).write(approve);
+    }
+
+    #[aztec(public)]
+    fn mint_public(to: AztecAddress, token_id: Field) {
+        assert(storage.minters.at(context.msg_sender()).read(), "caller is not minter");
+        // assert the token_id does not exist yet and mint
+        ...
+    }
+
+    #[aztec(public)]
+    fn mint_private(token_id: Field, secret_hash: Field) {
+        assert(storage.minters.at(context.msg_sender()).read(), "caller is not minter");
+        let pending_shields = storage.pending_shields;
+        let mut note = TransparentNote::new(token_id, secret_hash);
+        pending_shields.insert_from_public(&mut note);
+        // set that the nft exists
+        ...
+    }
+
+    #[aztec(public)]
+    #[aztec(view)]
+    fn public_get_name() -> pub FieldCompressedString {
+        storage.name.read_public()
+    }
+
+    #[aztec(private)]
+    #[aztec(view)]
+    fn private_get_name() -> pub FieldCompressedString {
+        storage.name.read_private()
+    }
+
+    #[aztec(public)]
+    #[aztec(view)]
+    fn public_get_symbol() -> pub FieldCompressedString {
+        storage.symbol.read_public()
+    }
+
+    #[aztec(private)]
+    #[aztec(view)]
+    fn private_get_symbol() -> pub FieldCompressedString {
+        storage.symbol.read_private()
+    }
+
+    #[aztec(public)]
+    #[aztec(view)]
+    fn admin() -> Field {
+        storage.admin.read().to_field()
+    }
+
+    #[aztec(public)]
+    #[aztec(view)]
+    fn is_minter(minter: AztecAddress) -> bool {
+        storage.minters.at(minter).read()
+    }
+
+    // Returns the total number of NFTs a given address owns in public.
+    #[aztec(public)]
+    #[aztec(view)]
+    fn balance_of_public(owner: AztecAddress) -> Field {
+        storage.public_balances.at(owner).read().to_integer()
+    }
+
+    #[aztec(public)]
+    fn transfer_public(from: AztecAddress, to: AztecAddress, token_id: Field, nonce: Field) {
+        if (!from.eq(context.msg_sender())) {
+            assert_current_call_valid_authwit_public(&mut context, from);
+        } else {
+            assert(nonce == 0, "invalid nonce");
+        }
+
+        ...
+    }
+
+    #[aztec(public)]
+    fn shield(from: AztecAddress, token_id: Field, secret_hash: Field, nonce: Field) {
```

Comment:
Decided to use the same shielding flow as standard token here to not complicate it now with different shield flows.